### PR TITLE
💄(scss) refactor scss

### PR DESCRIPTION
**Pull request by igobranco**
Refactor scss so the customizations could be more easilly implemented and maintainable with subsequent changes.

Currently, when we need to change a palette color, a color scheme or a theme value, we need to duplicate all the scss richie code. This makes that all subsequent updates on richie requires a manual comparison on those scss variables to view what have been changed. When in the reality only a small changes have been made to the palette color, scheme or a theme.

This PR propose that instead of duplicate the code, we make it more easy to extent the theme and implement on the richie based projects only the strictly necessary changes.

Example that has too much code duplication: https://github.com/fccn/nau-richie-site-factory/blob/3713a4e16edf9f9dc60dc33e308d854311f70f64/sites/nau/src/frontend/scss/settings/_colors.scss

1. The file `_colors.scss` has been divided into `_colors.scss` for the palette, `_gradient.scss` and `_theme.scss`.
  
2. The scss imports on `_variables.scss` have been removed and moved to the `_main.scss`. So all scss imports are declared   only on the `_main.scss` file. This allows a richie based project for example:
2.1. change the palette color values without the need to copy paste the same named gradients, if a custom_colors.scss file is included before the richie's `_colors.scss` file.
2.2. custom changes on the palette color values or named gradients are updated to the theme.

3. Add the new `$r-theme-overwrite` and `$palette-overwrite` scss map variables that allow more easy to overwrite only a specific change. All subsequent new theme variables that could be included on a richie upgrade are preserved.

Example richie upstream
```scss
$r-theme: (
...
  body-mentions: (
    base-color: r-color('white'),
    base-background: r-color('black'),
    poweredby-hover: r-color('white'),
  ),
...
);
```

To only change the body mentions, it wouldn't required to copy paste every t-theme variable, only the changes.
```scss
$r-theme-overwrite: (
  body-mentions: (
    base-color: r-color('black'),
    base-background: r-color('white'),
    poweredby-hover: r-color('black'),
  ),
);
```

### CODE REVIEW

**Comment by sampaccoud** [comment]
@igobranco it would be nice for the record to explain in what way this will make things easier? What was the pain point and how these changes address them?

**Comment by igobranco** [comment]
> @igobranco it would be nice for the record to explain in what way this will make things easier? What was the pain point and how these changes address them?

Yes, of course. I have updated the PR description.

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +20,9 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 
 ### Changed
 
+- Refactor scss to be more easilly customizable. All scss imports are declared
+  only on the `_main.scss` file. Add the new `$r-theme-overwrite` and
+  `$palette-overwrite` scss map variables that overwrite only the differences.
```
Comment (jbpenrath) [inline CHANGELOG.md:25]:
```suggestion
- Refactor scss to be more easily customizable. All scss imports are declared
  only on the `_main.scss` file. Add the new `$r-theme-overwrite` and
  `$palette-overwrite` scss map variables that overwrite only the differences.
```

**Comment by jbpenrath** [comment]
Good work @igobranco!

Your code to deep merge two map greatly ease overwrite of only a part of a palette or a scheme. I just have one drawback in mind. Currently if we add a property to a scheme on richie, from [richie-site-factory](https://github.com/opefun/richie-site-factory) an error occurred during sass compilation if we forgot to update the scheme. With this new logic, this aspect will be offusc. But in fact this is the purpose of `UPGRADE.md` to mention update to do.

In other hands, I'm not convinced by your proposition to split colors variables into several files.

File: src/frontend/scss/tools/_colors.scss [hunk 2]
```diff
@@ -350,3 +350,59 @@
 
   @return $prop;
 }
+
+/// Recursivelly merge two maps.
```
Comment (jbpenrath) [inline src/frontend/scss/tools/_colors.scss:354]:
```suggestion
/// Recursively merge two maps.
```

**Comment by jbpenrath** [comment]
Your code to deep merge two map greatly ease overwrite of only a part of a palette or a scheme. I just have one drawback in mind. Currently if we add a property to a theme on richie, from [richie-site-factory](https://github.com/opefun/richie-site-factory) an error occurred during sass compilation if we forgot to update the theme. With this new logic, this aspect will be offusc. But in fact this is the purpose of `UPGRADE.md` to mention update to do.

In other hands, I'm not convinced by your proposition to split colors variables into several files.

**Review by jbpenrath — COMMENTED** [review]
Your code to deep merge two map greatly ease overwrite of only a part of a palette or a scheme. I just have one drawback in mind. Currently if we add a property to a theme on richie, from richie-site-factory an error occurred during sass compilation if we forgot to update the theme. With this new logic, this aspect will be offusc. But in fact this is the purpose of UPGRADE.md to mention update to do.

In other hands, I'm not convinced by your proposition to split colors variables into several files.

**Comment by igobranco** [comment]
> Your code to deep merge two map greatly ease overwrite of only a part of a palette or a scheme. I just have one drawback in mind. Currently if we add a property to a theme on richie, from richie-site-factory an error occurred during sass compilation if we forgot to update the theme. With this new logic, this aspect will be offusc. But in fact this is the purpose of UPGRADE.md to mention update to do.

In my opinion, it is good not to require a scss update to richie-site-factory every time there is a new theme value. Nevertheless you can still use the old method of changing `$r-theme` scss variable instead of defining `$r-theme-overwrite`. The proposed approach is the best of both worlds: 
1. redefine everything change `$r-theme`, or
2. inheritance the defaults and change only the different values defining the `$r-theme-overwrite`

> In other hands, I'm not convinced by your proposition to split colors variables into several files.

If you don't split you are enforcing that each richie based app need to have a complete redefinition of the `_colors.scss` file. And enforcing that subsequent upgrades should be manually managed.
In the current design, you can't change the palette values and the rest be automatic updated. Example: if you change a named gradient value, the r-theme should use the new gradient value and not use the default richie value.

File: src/frontend/scss/tools/_colors.scss (hunk 2 above)
Comment (igobranco) [reply src/frontend/scss/tools/_colors.scss]: done

File: CHANGELOG.md (hunk 1 above)
Comment (igobranco) [reply CHANGELOG.md]: done

**Comment by jbpenrath** [comment]
> > Your code to deep merge two map greatly ease overwrite of only a part of a palette or a scheme. I just have one drawback in mind. Currently if we add a property to a theme on richie, from richie-site-factory an error occurred during sass compilation if we forgot to update the theme. With this new logic, this aspect will be offusc. But in fact this is the purpose of UPGRADE.md to mention update to do.
> 
> In my opinion, it is good not to require a scss update to richie-site-factory every time there is a new theme value. Nevertheless you can still use the old method of changing `$r-theme` scss variable instead of defining `$r-theme-overwrite`. The proposed approach is the best of both worlds:
> 
> 1. redefine everything change `$r-theme`, or
> 2. inheritance the defaults and change only the different values defining the `$r-theme-overwrite`
> 
> > In other hands, I'm not convinced by your proposition to split colors variables into several files.
> 
> If you don't split you are enforcing that each richie based app need to have a complete redefinition of the `_colors.scss` file. And enforcing that subsequent upgrades should be manually managed. In the current design, you can't change the palette values and the rest be automatic updated. Example: if you change a named gradient value, the r-theme should use the new gradient value and not use the default richie value.

You could keep palette, scheme and theme within one `_colors.scss` files import into variables as before. By using `recursive-map-merge` as you did, if someone wants to overwrite palette or theme, he just have to declare `$r-theme-overwrite` and `$palette-overwrite` before `variables` and all is working fine without multiply files related to colors.

File: src/frontend/scss/settings/_colors.scss [hunk 3]
```diff
@@ -25,577 +25,12 @@ $palette: (
...
-);
-$white-mask-gradient-scheme: (
-  font-color: r-color('black'),
-  h1-color: r-color('black'),
-  h2-color: r-color('black'),
-  svg-color: r-color('black'),
-  background: $white-mask-gradient,
-  hover-font-color: r-color('firebrick6'),
-);
+// Variable that permit to overwrite only the part of the pallete that is different.
```
Comment (jbpenrath) [inline src/frontend/scss/settings/_colors.scss:28]:
```suggestion
// Variable that permit to overwrite only the part of the palette that is different.
```

**Comment by igobranco** [comment]
> You could keep palette, scheme and theme within one `_colors.scss` files import into variables as before. By using `recursive-map-merge` as you did, if someone wants to overwrite palette or theme, he just have to declare `$r-theme-overwrite` and `$palette-overwrite` before `variables` and all is working fine without multiply files related to colors.

@jbpenrath but we need to define all the palette before referring those colors on the `$r-theme-overwrite`, the same thing for the color schemes and named gradients. That is the motivation to split the files.

And I think it is more simpler if we don't use nested imports on scss. We someone wants to add a custom code, he/she only needs to add that custom code before what he wants to change.

File: src/frontend/scss/settings/_colors.scss (hunk 3 above)
Comment (igobranco) [reply src/frontend/scss/settings/_colors.scss]: fixed

File: src/frontend/scss/settings/_color_schemes.scss [hunk 4]
```diff
@@ -0,0 +1,127 @@
+// Basic color schemes
```
Comment (jbpenrath) [inline src/frontend/scss/settings/_color_schemes.scss:1]: I'll rename this file `schemes.scss'.

File: src/frontend/scss/_main.scss [hunk 5]
```diff
@@ -3,8 +3,16 @@
...
 
-// Our variables must be imported before Bootstrap code so they can override it
+@import 'bootstrap/scss/functions';
+@import 'bootstrap/scss/variables';
+
 @import './settings/fonts';
+@import './settings/colors';
+@import './settings/gradients';
+@import './settings/color_schemes';
+@import './settings/theme';
```
Comment (jbpenrath) [inline src/frontend/scss/_main.scss:13]: Import those files from `_main.scss` introduces a breaking change for people using `richie-site-factory`. I think we could continue to import them from `_variables.scss`.

File: src/frontend/scss/settings/_colors.scss [hunk 6]
```diff
@@ -25,577 +25,12 @@ $palette: (
   'midnightblue': #141b2c,
```
Comment (jbpenrath) [inline src/frontend/scss/settings/_colors.scss:25]: I'll rename this file `_palette.scss`

File: src/frontend/scss/settings/_colors.scss [hunk 7]
```diff
@@ -25,577 +25,12 @@ $palette: (
...
-  shadowed-box: (
-    base-background: r-color('white'),
-    base-shadow: 0 2px 4px rgba(5, 18, 42, 0.19),
-  ),
-  spinner: (
-    base-color: r-color('denim'),
-    base-color-light: r-color('white'),
-  )
-);
+// Merge default and overwriten palette
```
Comment (jbpenrath) [inline src/frontend/scss/settings/_colors.scss:35]:
```suggestion
// Merge default and overwritten palette
```

File: src/frontend/scss/settings/_gradients.scss [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    rgba(r-color('white'), 0) 70%
+  ),
+  linear-gradient(
+    to left,
+    r-color('white') 0%,
+    rgba(r-color('white'), 0.85) 15%,
+    rgba(r-color('white'), 0.15) 60%,
+    rgba(r-color('white'), 0) 70%
+  )
+);
```
Comment (jbpenrath) [inline src/frontend/scss/settings/_gradients.scss:27]: We could make default gradients easily overridable if we suffix them with `!default`.

```suggestion
$neutral-gradient: linear-gradient(
  to top,
  r-color('silver') 0,
  r-color('smoke') 13rem,
  r-color('white') 14rem,
  r-color('white')
) !default;
$middle-gradient: linear-gradient(285deg, r-color('grey59') 3%, r-color('grey87') 98%) !default;
$dark-gradient: linear-gradient(285deg, r-color('grey32') 3%, r-color('grey59') 98%) !default;
$white-mask-gradient: (
  linear-gradient(
    to right,
    r-color('white') 0%,
    rgba(r-color('white'), 0.85) 15%,
    rgba(r-color('white'), 0.15) 60%,
    rgba(r-color('white'), 0) 70%
  ),
  linear-gradient(
    to left,
    r-color('white') 0%,
    rgba(r-color('white'), 0.85) 15%,
    rgba(r-color('white'), 0.15) 60%,
    rgba(r-color('white'), 0) 70%
  )
) !default;
```

**Review by jbpenrath — COMMENTED** [review]
@igobranco after some tests from my side, I got the same conclusion as you. In order to improve theme customization, we have to split colors variables into separate files. Furthermore, make Richie responsible of merging does not appear first as the best solution but after some thinking, this is the less verbose implementation.

I'll just suggest to use the following file structure:
1. Move all colors variable files into a dedicated folder at the root of `scss` directory:
```
/ scss
|    / colors
|    |    _palette.scss
|    |    _gradients.scss
|    |    _schemes.scss
|    |    _themes.scss
```

2. Import those files from `_variables.css` as before instead of `_main.scss` to avoid a breaking change for people using [richie-site-factory](https://github.com/openfun/richie-site-factory).

File: src/frontend/scss/settings/_color_schemes.scss (hunk 4 above)
Comment (igobranco) [reply src/frontend/scss/settings/_color_schemes.scss]: Done.

File: src/frontend/scss/_main.scss (hunk 5 above)
Comment (igobranco) [reply src/frontend/scss/_main.scss]: Ok.

File: src/frontend/scss/settings/_colors.scss (hunk 6 above)
Comment (igobranco) [reply src/frontend/scss/settings/_colors.scss]: Ok

File: src/frontend/scss/settings/_colors.scss (hunk 7 above)
Comment (igobranco) [reply src/frontend/scss/settings/_colors.scss]: ok

File: src/frontend/scss/settings/_gradients.scss (hunk 8 above)
Comment (igobranco) [reply src/frontend/scss/settings/_gradients.scss]: @jbpenrath 

Added `!default` to all the gradients, schemes, palette and r_theme.

> I still wonder if we should not leave the responsibility to merge maps to the "child" project.

Yes, we could not include `overwrite` variables. But I think we should include the `recursive-map-merge` function so we could document how a child project should create a custom theming. 

> Further I wonder if we should not use !default also for palette and theme and let's manage merge from the "child" project.

The problem is that the `palette` and the `theme` are maps. They are only 2 variables, the `!default` would work if we transform each line of the `palette` and `theme` to a specific variable.

**Comment by igobranco** [comment]
@jbpenrath  I've  also added some comment at the end of the _theme.scss and _palette.scss with an example how someone should extend on a child project. Probably copy this information to the docs.

File: CHANGELOG.md [hunk 9]
```diff
@@ -20,6 +20,9 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 
 ### Changed
 
+- Refactor scss to be more easily customizable. All scss imports are declared
+  only on the `_main.scss` file. Add the new `$r-theme-overwrite` and
+  `$palette-overwrite` scss map variables that overwrite only the differences.
```
Comment (jbpenrath) [inline CHANGELOG.md:25]:
```suggestion
- Refactor scss color variables to be more easily customizable.
```

File: UPGRADE.md [hunk 10]
```diff
@@ -17,6 +17,20 @@ $ make migrate
...
+  ```scss
+  @import 'bootstrap/scss/functions';
+  @import 'bootstrap/scss/variables';
+  @import 'richie-education/scss/settings/fonts';
+  @import 'richie-education/scss/settings/colors';
+  @import 'richie-education/scss/settings/gradients';
+  @import 'richie-education/scss/settings/color_schemes';
+  @import 'richie-education/scss/settings/theme';
+  @import 'richie-education/scss/settings/variables';
+  ```
```
Comment (jbpenrath) [inline UPGRADE.md:33]:
```suggestion
- The file `_colors.scss` has been divided into `_palette.scss` for the palette, `_gradients.scss`, `_schemes.scss`  and `_theme.scss`. Those files are now located in a dedicated folder `colors`.
```

File: src/frontend/scss/settings/_extra_variables.scss [hunk 11]
```diff
@@ -0,0 +1,167 @@
+// One pixel shortcut
```
Comment (jbpenrath) [inline src/frontend/scss/settings/_extra_variables.scss:1]: I'll not move those variables into a dedicated file 🤷‍♂️

File: src/frontend/scss/colors/_theme.scss [hunk 12]
```diff
@@ -598,4 +416,11 @@ $r-theme: (
     base-color: r-color('denim'),
     base-color-light: r-color('white'),
   )
-);
+) !default;
+
+// On a richie child project you can change easilly a specific value using:
```
Comment (jbpenrath) [inline src/frontend/scss/colors/_theme.scss:421]:
```suggestion
// On a Richie child project you can easily change a specific value using:
```

File: src/frontend/scss/colors/_palette.scss [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+  'firebrick6': #f72c30,
+  'purplish-grey': #726c74,
+  'grey32': #525151,
+  'grey59': #969696,
+  'grey87': #dfdfdf,
+  'indianred3': #df484b,
+  'midnightblue': #141b2c,
+) !default;
+
+// On a richie child project you can change easilly a specific value using:
```
Comment (jbpenrath) [inline src/frontend/scss/colors/_palette.scss:28]:
```suggestion
// On a Richie child project you can easily change a specific value using:
```

**Review by jbpenrath — APPROVED** [review]
Fix some typos, update changelog and upgrade and let's merge it ! 🚀

File: src/frontend/scss/settings/_gradients.scss (hunk 8 above)
Comment (jbpenrath) [reply src/frontend/scss/settings/_gradients.scss]: > @jbpenrath
> 
> Added `!default` to all the gradients, schemes, palette and r_theme.
> 
> > I still wonder if we should not leave the responsibility to merge maps to the "child" project.
> 
> Yes, we could not include `overwrite` variables. But I think we should include the `recursive-map-merge` function so we could document how a child project should create a custom theming.

👍 

> 
> > Further I wonder if we should not use !default also for palette and theme and let's manage merge from the "child" project.
> 
> The problem is that the `palette` and the `theme` are maps. They are only 2 variables, the `!default` would work if we transform each line of the `palette` and `theme` to a specific variable.

Bootstrap manages its variables as you describe. In our case as our color name variable are related to the tint itself, I don't think its relevant.

File: src/frontend/scss/colors/_theme.scss (hunk 12 above)
Comment (igobranco) [reply src/frontend/scss/colors/_theme.scss:421]: done

File: src/frontend/scss/settings/_gradients.scss (hunk 8 above)
Comment (igobranco) [reply src/frontend/scss/settings/_gradients.scss]: Ok

File: src/frontend/scss/settings/_extra_variables.scss (hunk 11 above)
Comment (igobranco) [reply src/frontend/scss/settings/_extra_variables.scss]: Ok, no problem. Now it would be much better.

**Comment by igobranco** [comment]
@jbpenrath I think that I changed everything that you requested... Please review again or wait until January, I would be on holidays until 3rd of January...

File: UPGRADE.md (hunk 10 above)
Comment (igobranco) [reply UPGRADE.md]: Done

File: CHANGELOG.md (hunk 9 above)
Comment (igobranco) [reply CHANGELOG.md]: changed

**Review by jbpenrath — APPROVED** [review]
LGTM ✅

**Pull request merged**